src/Services/ServiceAccountJwtRetriever.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace KeycloakAuthGuard\Services;
+
+use Illuminate\Http\Client\RequestException;
+use Illuminate\Support\Facades\Config;
+use Illuminate\Support\Facades\Http;
+
+readonly class ServiceAccountJwtRetriever implements ServiceAccountJwtRetrieverInterface
+{
+    private string $keycloakBaseUrl;
+
+    private string $realm;
+
+    public function __construct(private string $clientId, private string $clientSecret)
+    {
+        $this->keycloakBaseUrl = trim(Config::get('keycloak.base_url'), '/');
+        $this->realm = Config::get('keycloak.realm');
+    }
+
+    /**
+     * @throws RequestException
+     */
+    public function getJwt(): string
+    {
+        return $this->getResponse()['access_token'];
+    }
+
+    /**
+     * @throws RequestException
+     */
+    public function getResponse(): array
+    {
+        return Http::asForm()->post($this->getJwtRetrieveUrl(), [
+            'grant_type' => 'client_credentials',
+            'client_id' => $this->clientId,
+            'client_secret' => $this->clientSecret,
+        ])->throw()->json();
+    }
```

Comment:
**[non-blocker]** I’m probably being over-cautious, but since we are sending the client secret with this request, doing it without TLS feels illegal. I would add this. But I’ll leave the decision to you, maybe I’m just paranoid. 😅 

```suggestion
    {
       if (!Str::startsWith($this->getJwtRetrieveUrl(), 'https')) {
           throw new InvalidConfigurationException("URL for requesting service account token is not secure (https): $this->getJwtRetrieveUrl()");
       }
    
        return Http::asForm()->post($this->getJwtRetrieveUrl(), [
            'grant_type' => 'client_credentials',
            'client_id' => $this->clientId,
            'client_secret' => $this->clientSecret,
        ])->throw()->json();
    }
```

Reply:
I wouldn't do this exact change as suggested:

- Running keycloak locally with just HTTP wouldn't be compatible with this change
- This is the only place where HTTPS requirement is stated, other parts will use HTTP connection

Instead I would add such requirement as global to the keycloak url when reading it from env vars. And such requirement should not be applied when running locally (APP_ENV) can be used for this

And from keycloak's perspective - it's keycloak's responsibility to not provide tokens when HTTPS is not used, which is currently configured as such :)

So I wouldn't do this change atm